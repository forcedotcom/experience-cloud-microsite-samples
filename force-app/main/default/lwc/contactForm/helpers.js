/**
 * Utility to handle redirect action
 *
 * @param {string} url Redirect url
 */
export function redirectPage(url) {
    window.open(url, '_self');
}

/**
 * Utility to verify form data
 *
 * @param {Array} inputs Lightning inputs from the form
 * @returns {Boolean} To show whether all inputs are valid
 */
export function isSubmitActionDataValid(inputs) {
    return inputs.every((input) => input?.validity?.valid);
}

/**
 * Utility to get form data
 *
 * @param {Array} inputs Lightning inputs from the form
 * @returns {Object} Object mapping fields to values
 */
export function transformFormData(inputs) {
    const fields = {};
    inputs.forEach((element) => {
        fields[element.name] = element.value;
    });
    return fields;
}

/**
 * Utility to format form data for contact form submission
 *
 * @param {object} fieldsData This is an object representing form fields. Eg. {FirstName: John, LastName: Doe}
 * @returns {object} Data formatted correctly and ready for insertion to marketing cloud
 */
export function transformFieldsForContactFormSubmission(fieldsData) {
    const formFields = Object.keys(fieldsData ?? {}).map((fieldName) => {
        const fieldValue = normalizeValue(fieldsData[fieldName]);
        return {
            name: fieldName,
            value: fieldValue
        };
    });
    return {
        formFieldsList: {
            formFields
        }
    };
}

/**
 * Utility to normalize all field values to text (all contact form fields are currently Text or Email)
 *
 * @param {string|boolean} value The form field to normalize
 * @returns {string} The string representation of the field
 */
export function normalizeValue(value) {
    if (value === null || value === undefined) {
        return '';
    }
    return String(value);
}

/**
 * Utility to check validity of contact form data
 *
 * @param {object} formInput The input for the contact Form
 * @returns {boolean} Is the form Input Data valid ?
 */
export function validateFormData(formInput) {
    const formFields = formInput?.formFieldsList?.formFields;
    return Array.isArray(formFields);
}
