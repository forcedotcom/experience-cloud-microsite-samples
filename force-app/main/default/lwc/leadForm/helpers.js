/**
 * Utility to handle redirect action
 *
 * @param {string} url Redirect url
 */
export function redirectPage(url) {
    window.open(url, '_self');
}

/**
 * Utility to verify form data
 *
 * @param {Array} inputs Lightning inputs from the form
 * @returns {Boolean} To show whether all inputs are valid
 */
export function isSubmitActionDataValid(inputs) {
    const invalids = inputs?.filter((element) => !element.validity.valid);

    if (invalids?.length > 0) {
        invalids.forEach((element) => element.reportValidity());
        return false;
    }
    return true;
}

/**
 * Utility to get form data
 *
 * @param {Array} inputs Lightning inputs from the form
 * @returns {Object} Object mapping Lead fields with values
 */
export function getFormData(inputs) {
    const fields = {};
    inputs.forEach((element) => {
        fields[element.name] =
            element.type === 'checkbox' ? element.checked : element.value;
    });
    return fields;
}
