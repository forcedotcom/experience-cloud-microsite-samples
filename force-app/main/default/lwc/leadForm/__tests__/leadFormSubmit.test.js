import { createElement } from 'lwc';
import LeadForm from 'c/leadForm';
import { createRecord } from 'lightning/uiRecordApi';
import * as helper from '../helpers';
import { DEFAULT_LEAD_FORM_PROPERTIES } from './testUtils';

/**
 * @param {object} props object with all properties for lead form component
 * @returns {object} - HTML DOM object
 */
function createComponent(props = {}) {
    const element = createElement('c-lead-form', {
        is: LeadForm
    });
    Object.assign(element, props);
    document.body.appendChild(element);
    return element;
}

/**
 * Helper sets required fields with default values
 *
 * @param {object} element - HTML DOM object reference
 */
function setDefaultValues(element) {
    //Set company default value
    const companyInput = element.shadowRoot.querySelector(
        'lightning-input[data-company]'
    );
    companyInput.value = 'Salesforce';

    //Set lastname default value
    const lastnameInput = element.shadowRoot.querySelector(
        'lightning-input[data-lastname]'
    );
    lastnameInput.value = 'Smith';
}

describe('Lead Form Component Submit Tests', () => {
    afterEach(() => {
        // The jsdom instance is shared across test cases in a single file so reset the DOM
        while (document.body.firstChild) {
            document.body.removeChild(document.body.firstChild);
        }
        // Prevent data saved on mocks from leaking between tests
        jest.clearAllMocks();
    });
    it('should show success message on inline action submission', async () => {
        //Create component and set required fields with default values
        const element = createComponent(DEFAULT_LEAD_FORM_PROPERTIES);
        setDefaultValues(element);

        //Mocking isSubmitActionDataValid
        const spy = jest.spyOn(helper, 'isSubmitActionDataValid');
        spy.mockImplementation(() => {
            return true;
        });

        //Initiate submit logic
        const form = element.shadowRoot.querySelector('form');
        form.dispatchEvent(new CustomEvent('submit'));

        //Resolve promise for API call
        await Promise.resolve();
        //Need to wait for then on API call
        await Promise.resolve();
        //Need to wait for then on the API callout logic
        await Promise.resolve();
        //Need to wait for flags to be changed in order to update UI
        await Promise.resolve();

        //Check if success message is present
        const pTag = element.shadowRoot.querySelector('p');
        expect(pTag.textContent).toEqual(
            DEFAULT_LEAD_FORM_PROPERTIES.submitActionSuccessMessage
        );
    });

    it('should open new window for redirect action', async () => {
        //Set property values
        DEFAULT_LEAD_FORM_PROPERTIES.submitActionType =
            'Redirect to existing page';

        //Create component and set required fields with default values
        const element = createComponent(DEFAULT_LEAD_FORM_PROPERTIES);
        setDefaultValues(element);

        //Mocking isSubmitActionDataValid
        const spy = jest.spyOn(helper, 'isSubmitActionDataValid');
        spy.mockImplementation(() => {
            return true;
        });

        //Creating jest function to mock window.open
        window.open = jest.fn().mockImplementation(() => true);

        //Initiate submit logic
        const form = element.shadowRoot.querySelector('form');
        form.dispatchEvent(new CustomEvent('submit'));

        //Need to resolve promise for API call
        await Promise.resolve();
        //Need to wait for then on API call
        await Promise.resolve();
        //Need to wait for window logic to complete
        await Promise.resolve();

        //Check if mock function has been clicked
        expect(window.open).toHaveBeenCalledTimes(1);
        //Check mock function parameters
        expect(window.open.mock.calls[0][0]).toEqual(
            DEFAULT_LEAD_FORM_PROPERTIES.redirectUrl
        );
    });

    it('should show error message createRecord api call throws an error', async () => {
        //Create component and set required fields with default values
        const element = createComponent(DEFAULT_LEAD_FORM_PROPERTIES);
        setDefaultValues(element);

        //Mocking isSubmitActionDataValid
        const spy = jest.spyOn(helper, 'isSubmitActionDataValid');
        spy.mockImplementation(() => {
            return true;
        });
        //Mock createRecord throwing an error
        createRecord.mockRejectedValue(new Error('Error!'));

        //Initiate the submit
        const form = element.shadowRoot.querySelector('form');
        form.dispatchEvent(new CustomEvent('submit'));

        //Need to resolve promise for API call
        await Promise.resolve();
        //Need to wait for then on API call
        await Promise.resolve();
        //Need to wait for then on the API callout logic
        await Promise.resolve();
        //Need to wait for flags to be changed in order to update UI
        await Promise.resolve();
        //Need to wait longer to apply changes
        await Promise.resolve();

        //Check if error message is present
        const pTag = element.shadowRoot.querySelector('p');
        expect(pTag.textContent).toEqual(
            DEFAULT_LEAD_FORM_PROPERTIES.submitActionErrorMessage
        );
    });

    it('should not fire submit event when fields have invalid inputs', async () => {
        const element = createComponent(DEFAULT_LEAD_FORM_PROPERTIES);
        expect(element).toMatchSnapshot();

        const submitHandler = jest.fn();
        element.addEventListener('submit', submitHandler);

        //Initiate submit logic
        const form = element.shadowRoot.querySelector('form');
        form.dispatchEvent(new CustomEvent('submit'));

        await Promise.resolve();

        expect(submitHandler).not.toBeCalled();
    });

    it('is accessible with default properties', async () => {
        //Create lead form component
        const element = createComponent(DEFAULT_LEAD_FORM_PROPERTIES);
        await expect(element).toBeAccessible();
    });

    it('is accessible with custom properties', async () => {
        //Create lead form component
        const element = createComponent(DEFAULT_LEAD_FORM_PROPERTIES);
        await expect(element).toBeAccessible();
    });

    it('is accessible in success state', async () => {
        //Create component and set required fields with default values
        const element = createComponent(DEFAULT_LEAD_FORM_PROPERTIES);
        setDefaultValues(element);

        //Mocking isSubmitActionDataValid
        const spy = jest.spyOn(helper, 'isSubmitActionDataValid');
        spy.mockImplementation(() => {
            return true;
        });
        //Initiate submit logic
        const form = element.shadowRoot.querySelector('form');
        form.dispatchEvent(new CustomEvent('submit'));

        //Resolve promise for API call
        await Promise.resolve();
        //Need to wait for then on API call
        await Promise.resolve();
        //Need to wait for then on the API callout logic
        await Promise.resolve();
        //Need to wait for flags to be changed in order to update UI
        await Promise.resolve();

        await expect(element).toBeAccessible();
    });

    it('is accessible in error state', async () => {
        //Create component and set required fields with default values
        const element = createComponent(DEFAULT_LEAD_FORM_PROPERTIES);
        setDefaultValues(element);

        //Mocking isSubmitActionDataValid
        const spy = jest.spyOn(helper, 'isSubmitActionDataValid');
        spy.mockImplementation(() => {
            return true;
        });

        //Mock createRecord implementation with bad response
        createRecord.mockImplementation(() => {
            return Promise.resolve(false);
        });

        //Initiate the submit
        const form = element.shadowRoot.querySelector('form');
        form.dispatchEvent(new CustomEvent('submit'));

        //Need to resolve promise for API call
        await Promise.resolve();
        //Need to wait for then on API call
        await Promise.resolve();
        //Need to wait for then on the API callout logic
        await Promise.resolve();
        //Need to wait for flags to be changed in order to update UI
        await Promise.resolve();
        //Need to wait longer to apply changes
        await Promise.resolve();

        await expect(element).toBeAccessible();
    });
});
