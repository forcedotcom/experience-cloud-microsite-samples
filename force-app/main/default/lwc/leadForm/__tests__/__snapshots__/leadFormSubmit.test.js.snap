// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lead Form Component Submit Tests should not fire submit event when fields have invalid inputs 1`] = `
<c-lead-form>
  #shadow-root(open)
    <form
      novalidate=""
    >
      <div
        class="slds-var-m-around_medium"
      >
        <lightning-input
          data-firstname="FirstName"
        >
          #shadow-root(open)
        </lightning-input>
        <lightning-input
          data-lastname="LastName"
        >
          #shadow-root(open)
        </lightning-input>
        <lightning-input
          data-email="Email"
        >
          #shadow-root(open)
        </lightning-input>
        <lightning-input
          data-phone="Phone"
        >
          #shadow-root(open)
        </lightning-input>
        <lightning-input
          data-company="Company"
        >
          #shadow-root(open)
        </lightning-input>
        <lightning-button>
          #shadow-root(open)
        </lightning-button>
      </div>
    </form>
</c-lead-form>
`;
